src/digest/digest.cr

```diff
@@ -133,13 +133,68 @@ abstract class Digest
     final dst
   end
 
+  # Writes the final digest output to `dst`.
+  #
+  # Faster than the `Bytes` allocating version.
+  # Use when hashing in loops.
+  #
+  # This method can only be called once and raises `FinalizedError` on subsequent calls.
+  #
+  # NOTE: `.dup.final(dst)` call may be used to get an intermediate hash value.
   def final(dst : Bytes) : Bytes
     check_finished
     @finished = true
     final_impl dst
     dst
   end
 
+  # Returns a hexadecimal-encoded digest in a new `String`.
+  #
+  # This method can only be called once and raises `FinalizedError` on subsequent calls.
+  #
+  # NOTE: `.dup.final` call may be used to get an intermediate hash value.
+  def hexfinal : String
+    sary = uninitialized StaticArray(UInt8, 64)
+    tmp = sary.to_slice[0, digest_size]
```

Comment:
Just wondering is there some effective limitation on `digest_size` that would ensure it's less than 64 bytes? Implementations could technically use whatever size they like. Practical implications might make digests bigger than 64 bytes unlikely, but I presume it would be possible?

If we can't excuse the case for `digest_size > 64` with good certainty, there should be an alternative here.

Reply:
Some newer hashes allow variable, almost infinite output sizes requiring each implementation to validate the size.

Either this:
```crystal
protected abstract def final_output_size_check(size) : Bool
````
or an overridable
```crystal
protected def final_output_size_check(size) : Bool
  true
end
```

Reply:
Sorry, I responded when busy.  

 Is there an equivalent to `ALLOCA` to dynamically stack alloc a `Bytes` compatible type?

Reply:
I don't think so.

We'll have to allocate on the heap when `digest_size > sary.size`.

Reply:
Or maybe a crazier alternative? In order to avoid an additional heap allocation we could re-purpose the the string buffer which already needs to be allocated for building the string (in `Slice#hexstring`).

```cr
string_size = digest_size * 2
string = String.new(string_size) do |buffer|
  tmp = Slice.new(buffer + digest_size, digest_size)
  final tmp
  tmp.hexstring(buffer)
  {string_size, string_size}
end

Reply:
If not I'd say leave it to those implementations to override `#hexfinal` or fix the issue in a future PR.

Reply:
That makes a lot of sense 👍 
We should add a note in the docs for `digest_size`.

Reply:
Output size limitations removed except for `hexfinal(io)`